added tests to VS2019 solution

LowLevelMahn opened the pull request:
-adds a new vs folder under tests with a VS2019 project
-builds with x64 Debug/Release (also HeavyDebug)
-currently all tests (also example and fs_utils) are in the same exe
-renamed all test to _tests.cpp because i think that the different folder name cpp/objs thing is untypical for VStudio development - and one can easier differentiate between tests and tested code
-meson build is not tested (could fail due to cpp file name changes) - currently too low hard disk space for another VM :/
-the STL asserts happening - the rwqueue and soft_limiter tests appeard after i've renamed the cpps to get unique obj files
-some tests failing currently

install gtest using vcpkg:
vcpkg install --triplet x64-windows gtest
vcpkg integrate install

hope its a good start

kcgen commented on tests/stubs.cpp:
```diff
@@ -30,3 +30,10 @@ Config *control = nullptr;
+const char *MSG_Get(char const *)
+{
+	return nullptr;
+};
```
Can you add a carriage return at the end? 

GitHub flags these with the little red circle - 
![2021-05-26_15-22](https://user-images.githubusercontent.com/1557255/119738802-21d43980-be36-11eb-9f5d-b2de15526857.png)

LowLevelMahn replied:
will do

kcgen commented on tests/vs/tests.vcxproj:
```diff
@@ -0,0 +1,167 @@
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
+  <ImportGroup Label="ExtensionTargets">
+  </ImportGroup>
+</Project>
```
(carriage return)

LowLevelMahn replied:
this file was generated by vstudio, but i can add a newline, i dont know if vstudio will keep it

kcgen replied:
I see VS dropped the CR again :-) 
Let's leave it knocked out.. if that's the way the tool prefers it, then so be it!

kcgen commented on tests/vs/tests.vcxproj.filters:
```diff
@@ -0,0 +1,64 @@
+  <ItemGroup>
+    <None Include="..\meson.build" />
+  </ItemGroup>
+</Project>
```
Here too.

kcgen reviewed (changes requested):
> hope its a good start

I'd have to say that an understatement :smile:  You've got unit tests fully working under Visual Studio!

Tested and working on my side:

![2021-05-26_15-12](https://user-images.githubusercontent.com/1557255/119738452-98bd0280-be35-11eb-811e-a5454d445122.png)

I also ran the code analysis, which provided some nice tips as well for more fixes and improvements.

kcgen commented:
The last comment is to use present tense or the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood), which is the style used for all the staging commits (and for git in general).

_"added tests to VS2019 solution"_

I ask myself _"This commit will .. "_ and then I finish the sentence as the commit message. 
In this case "add unit tests to the VS2019 solution" (or something like that :-)

kcgen commented:
To rename the commit wording, you can use the "amend" feature.  In the Visual Studio Team Explorer pane, go to Sync, lookup the latest Outgoing Commit edit the message and choose “amend” from the options. Or type `git commit –amend -m “My new message”` on the command line.

If you add a new commit to address the comments here, you can "squash" it into your original commit using the very nice notes here: https://medium.com/corrado-cavalli/squash-commits-in-visual-studio-b469ef021436

LowLevelMahn commented:
> To rename the commit wording, you can use the "amend" feature. In the Visual Studio Team Explorer pane, go to Sync, lookup the latest Outgoing Commit edit the message and choose “amend” from the options.
> 
> Or type `git commit --amend -m "My new message"` on the command line.
> 
> If you add a new commit to address the comments here, you can "squash" it into your original commit using the notes here: https://medium.com/corrado-cavalli/squash-commits-in-visual-studio-b469ef021436

i've added the new commit message and squashed the commits into one - everything seems fine localy
but now i can't fetch&merge my github repo with staging-master anymore?

![image](https://user-images.githubusercontent.com/580819/119768645-3e658580-beb9-11eb-9fe1-fb510da6716c.png)

kcgen commented:
After squashing, your local history should show just one commit with the new commit message and the couple new carriage returns all in the same commit, plus all your original work (all in the same commit). 

Run `git log` to show your local history, just to confirm as a double check again the visual studio graphical output. 

Once you're confirmed down to one commit, then `git push -f`, and that will overwrite your remote history with your new cleaned up single commit. At that point, your local and remotes will match eachother and will all be in-sync. 

That will also propagate automatically upstream to your PR.

LowLevelMahn commented:
> After squashing, your local history should show just one commit with the new commit message and the couple new carriage returns all in the same commit, plus all your original work (all in the same commit).
> 
> Run `git log` to show your local history, just to confirm as a double check again the visual studio graphical output.
> 
> Once you're confirmed down to one commit, then `git push -f`, and that will overwrite your remote history with your new cleaned up single commit. At that point, your local and remotes will match eachother and will all be in-sync.
> 
> That will also propagate automatically upstream to your PR.

i'll do my best - also currently fixing x86 build

kcgen commented:
Also, see lines 37 and 47 in `fs_utils_tests.cpp`, which should be updated with the renamed filename.

kcgen commented:
If the git squashing isn't going well then tomorrow AM my time I can help more on the git side, and give exact steps. 
(To experiment with squashing, you can Zip your repo directory or take a backup, just incase something bad happens).

LowLevelMahn commented:
> If the git squashing isn't going well then tomorrow AM my time I can help more on the git side, and give exact steps.
> (To experiment with squashing, you can Zip your repo directory or take a backup, just incase something bad happens).

yes, i think i need help, with git and VS2019

i can't compile the under x64 - even if i remove my repo/re-clone and checkout the very first (squashed) commit

"add unit tests to the VS2019 solution"
26.05.2021 07:57:56
1e7ebfcf6e89b3b953616408e720827c410c0f0f

setup.cpp fails to build with x64 or x86

```
1>------ Neues Erstellen gestartet: Projekt: tests, Konfiguration: Debug x64 ------
1>cross.cpp
1>fs_utils_win32.cpp
1>rwqueue.cpp
1>setup.cpp
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(155,9): warning C4244: "Argument": Konvertierung von "Bits" in "int", möglicher Datenverlust
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(164,9): warning C4244: "=": Konvertierung von "Bits" in "int", möglicher Datenverlust
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(180,2): error C3861: "lowcase": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(234,2): error C3861: "assertm": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(251,2): error C3861: "upcase": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(258,2): error C3861: "upcase": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(347,33): error C3861: "lowcase": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(379,2): error C2653: "Cross": Keine Klasse oder Namespace
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(379,9): error C3861: "ResolveHomedir": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(382,39): error C2065: "CROSS_FILESPLIT": nichtdeklarierter Bezeichner
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(384,6): error C2653: "Cross": Keine Klasse oder Namespace
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(384,13): error C3861: "IsPathAbsolute": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(404,33): warning C4244: "Argument": Konvertierung von "Bitu" in "int", möglicher Datenverlust
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(681,2): error C3861: "trim": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(689,2): error C3861: "trim": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(690,2): error C3861: "trim": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(693,7): error C3861: "strcasecmp": Bezeichner wurde nicht gefunden.
1>D:\projects\fun\dos_games_rev\dosbox_dev\llm_dosbox-staging\src\misc\setup.cpp(714,50): warning C4267: "Argument": Konvertierung von "size_t" nach "const _Ty", Datenverlust möglich
```

i can fix part of the compile errors - but come into a dependency problem in the end

the only thing i did was:
1. updated VS2019 to 16.10, due to C++20 problems in another problem
2. installed the x86 libs with vcpkg (removing does not help)

i have no idea what changed?

kcgen commented:
I worked through the squash steps and wrote it up (sorry for the delay!) hopefully you can reproduce them exactly on your side.

First we need to set Git's preferred editor.  Let's use a [64-bit build of Notepad++](https://github.com/notepad-plus-plus/notepad-plus-plus/releases/download/v7.9.5/npp.7.9.5.Installer.x64.exe), 

On a normal `cmd.exe` shell, run:

`git config --global core.editor "'C:/Program Files/Notepad++/notepad++.exe' -multiInst -notabbar -nosession -noPlugin"`

Now change directories into your dosbox-staging repo, and run: `git config -e`. 

This should pop-open your editor, and you should see Git waiting on the command line until you exit the editor like this:
![2021-05-27_17-37](https://user-images.githubusercontent.com/1557255/119924863-37765b80-bf29-11eb-93c4-d0d8d7887a09.png)

We'll now squash the commits using Git's "rebase" feature. 

First, we need count how many commits "back in time" we want to operate on.  In this case we want to go back 8 commits, and squash them into the 9th:

![2021-05-27_16-51](https://user-images.githubusercontent.com/1557255/119924872-39401f00-bf29-11eb-889a-a5439e85c979.png)

Run `git rebase -i HEAD~7`, in the root of your repo (the ~ itself counts for "one back")

![2021-05-27_16-54](https://user-images.githubusercontent.com/1557255/119924871-39401f00-bf29-11eb-8a35-aec93bf831be.png)

This will open up a list of these commits, in reverse order, in you editor (I'm using `nano` on Linux here). You'll see this in Notepad++:

![2021-05-27_16-55](https://user-images.githubusercontent.com/1557255/119924870-38a78880-bf29-11eb-935d-042fcd06e0df.png)

There's more information further down in the text file - feel free to read it all.  This text file acts as our "rebase instructions" for git that it will perform once we save and exit.  

Replace the "pick" words with "s", like I've done here, for the bottom 6 commits:

![2021-05-27_16-56](https://user-images.githubusercontent.com/1557255/119924869-38a78880-bf29-11eb-9b44-ceb0e423f855.png)

Save and exit the file. Git will start squashing them, and then return you to the editor. This time, we're editing the commit message for the resulting "squashed commit".  In this case, we're given all of the commits messages on-by-one to work with:

![2021-05-27_17-19](https://user-images.githubusercontent.com/1557255/119924867-38a78880-bf29-11eb-8feb-d7b2945bb547.png)

I delete all the lines exist for the message I want, which is this: 

![2021-05-27_17-20](https://user-images.githubusercontent.com/1557255/119924866-38a78880-bf29-11eb-84d8-f5fe6166e9e2.png)

Save and exit the editor. 

Git then finishes and leaves you with the results:

![2021-05-27_17-20_1](https://user-images.githubusercontent.com/1557255/119924865-380ef200-bf29-11eb-8928-676540635b45.png)

Now do a `git push -f`, and there should now be a single combined commit sitting in the PR branch (ready to merge!).

LowLevelMahn commented:
> i can't compile the test anymore (x64) - even if i remove my repo/re-clone and checkout the very first (squashed) commit

do it still compile with your VS2019 version?

LowLevelMahn commented:
> I worked through the squash steps and wrote it up (sorry for the delay!) hopefully you can reproduce them exactly on your side.

thanks for the very detailed description, will try it now

kcgen commented:
> but i still don't understand why your PVS Meson commit is also on the pick list?

In the list of commits (before we squashed) two of them were 'merge commits', which in the text editor didn't show up.  It's likely that these didn't count toward the 7, and therefore it went "too far back". 

But there's little harm.. we could have selected 20 back and see all those older commits too; so long as they all stay as 'pick' then they will be left as-is.

LowLevelMahn commented:
is the gtest package missing for MSVC x86 build?

kcgen commented:
> is the gtest package missing for MSVC x86 build?

Yes - here's the workflow file:

https://github.com/dosbox-staging/dosbox-staging/actions/runs/884353766/workflow

It's part of the repo, inside `./github/workflow/windows.yml`, which you can edit and push. These are just basic commandline steps. 

Do you know how to limit the build to just the 'dosbox' solution, and exclude the tests?

Here's the command in the workflow file, which seems to be building the DOSBox binary plus the tests:

`MSBuild -m dosbox.sln -p:Configuration=Debug `

We ultimately do want to include the tests in the CI build.. but we know they will always fail right now until they're fixed.. so it also makes sense that they're off until then.

LowLevelMahn commented:
> > is the gtest package missing for MSVC x86 build?
> 
> Yes - here's the workflow file:
> 
> https://github.com/dosbox-staging/dosbox-staging/actions/runs/884353766/workflow
> 
> It's part of the repo, inside `./github/workflow/windows.yml`, which you can edit and push. These are just basic commandline steps.

i'll fix it

> Do you know how to limit the build to just the 'dosbox' solution, and exclude the tests?

i think that will work 

 `MSBuild -m dosbox.sln /target:dosbox -p:Configuration=Debug `

target is the project

> We ultimately do want to include the tests in the CI build.. but we know they will always fail right now until they're fixed.. so it also makes sense that they're off until then.

do mean the test-run or the build of it?

i currently can't build the test - due to my VS2019 update (i think) - boost 1.72 also does not compile anymore with the latest version

kcgen commented:
I forgot to answer you before, but yes, on my local systems, the unit tests still build (after the squashing).

> i think that will work `MSBuild -m dosbox.sln /target:dosbox -p:Configuration=Debug ... `

Excellent! Yes, I see `-target:<target>` mentioned here too:  https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2019

>  do mean the test-run or the build of it?

Yes to both: I think we should not even attempt to build the unit tests in the CI workflow until we have them build and running clean locally.   Once we do, we can push a PR that compiles the unit tests and also runs them (exactly like we're doing in the Linux workflow).

LowLevelMahn commented:
> until we have them build and running clean locally.

easy to say for an non VS2019 16.10 user :) 

there is no downgrade option - and reinstalling will directly install 16.10 - so i can't currently test it locally

kcgen commented:
> > until we have them build and running clean locally.
> 
> easy to say for an non VS2019 16.10 user :)
> 
> there is no downgrade option - and reinstalling will directly install 16.10 - so i can't currently test it locally

Ouch!  Thanks for this heads up and that link.. what a disaster.  Hopefully it's fixed soon. I will avoid "upgrading" for now.

Oh - the cache is only saved if the run passes (otherwise the cache could "self pollute" and ahrm subsequent runs).

Regarding your suggested `-target:dosbox` setting to build just to the dosbox binary: I think it makes sense to use this, given the overall `MSBuild ... ` output is piped to a log file, which is later fed to the warning counter.  If the warnings exceed the current allow limit then the run is considered to have failed (and I'm pretty sure the unit tests have a handful of their own warnings).

kcgen commented:
@LowLevelMahn - can you update the Windows workflow file again, and add the `-target:dosbox` settings to the MSBuild lines? That should get the build passing.

kcgen commented:
Let's merge this and open the door to unit tests on Windows! More coverage is a very good thing. 
Thank you, @LowLevelMahn.

LowLevelMahn commented:
> Ouch! Thanks for this heads up and that link.. what a disaster. Hopefully it's fixed soon. I will avoid "upgrading" for now.

happens also on Win10 when updating VStudio
